site/src/pages/WorkspacePage/WorkspaceScheduleControls.tsx

```diff
@@ -48,28 +44,50 @@ export const WorkspaceScheduleControls: FC<WorkspaceScheduleControlsProps> = ({
   );
   const deadlinePlusEnabled = maxDeadlineIncrease >= 1;
   const deadlineMinusEnabled = maxDeadlineDecrease >= 1;
-
-  const onDeadlineChangeSuccess = () => {
-    displaySuccess("Updated workspace shutdown time.");
-  };
-  const onDeadlineChangeFails = (error: unknown) => {
-    displayError(
-      getErrorMessage(error, "Failed to update workspace shutdown time."),
+  const deadlineUpdateTimeout = useRef<number>();
+  const lastStableDeadline = useRef<Dayjs>(deadline);
+
+  const updateWorkspaceDeadlineQueryData = (deadline: Dayjs) => {
+    queryClient.setQueryData(
+      workspaceByOwnerAndNameKey(workspace.owner_name, workspace.name),
+      {
+        ...workspace,
+        latest_build: {
+          ...workspace.latest_build,
+          deadline: deadline.toISOString(),
+        },
+      },
     );
   };
-  const decreaseMutation = useMutation({
-    ...decreaseDeadline(workspace),
-    onSuccess: onDeadlineChangeSuccess,
-    onError: onDeadlineChangeFails,
-  });
-  const increaseMutation = useMutation({
-    ...increaseDeadline(workspace),
-    onSuccess: onDeadlineChangeSuccess,
-    onError: onDeadlineChangeFails,
+
+  const updateDeadlineMutation = useMutation({
+    ...updateDeadline(workspace),
+    onSuccess: (_, updatedDeadline) => {
+      displaySuccess("Workspace shutdown time has been successfully updated.");
+      lastStableDeadline.current = updatedDeadline;
+    },
+    onError: (error) => {
+      displayError(
+        getErrorMessage(
+          error,
+          "We couldn't update your workspace shutdown time. Please try again.",
+        ),
+      );
+      updateWorkspaceDeadlineQueryData(lastStableDeadline.current);
```

Comment:
Is storing the previous deadline necessary here, or could we use the [rollback value](https://tanstack.com/query/v4/docs/framework/react/guides/optimistic-updates) from `useMutation`'s context?

Reply:
Hm... I will try to use it.

Reply:
Since I'm updating a nested field of the workspace - the "root data structure" - I would have to write something that IMO would add more complexity than what we have right now.